server/vulnerabilities/vuln_centos/centos.go

```diff
@@ -0,0 +1,512 @@
+			return
+		}
+		if nonReleasePathRegex.MatchString(path.Join(e.Request.URL.Path, href)) {
+			return
+		}
+		if !recentCentOSPathRegex.MatchString(path.Join(e.Request.URL.Path, href)) {
+			if verbose {
+				fmt.Printf("Ignoring old release: %s\n", path.Join(e.Request.URL.Path, href))
+			}
+			return
+		}
+		if href == "repomd.xml" {
+			u := *e.Request.URL
+			u.Path = path.Join(u.Path, href)
+			repoMDs = append(repoMDs, u)
+			if verbose {
+				fmt.Printf("%s\n", u.Path)
+			}
+			return
+		}
+		if !strings.Contains(href, "/") {
+			return
+		}
+		e.Request.Visit(href)
+	})
+
+	c.AllowedDomains = append(c.AllowedDomains, repositoryDomain)
+
+	if root == "" {
+		root = defaultRoot
+	}
+	if err := c.Visit(repositoryURL + root); err != nil {
+		return err
+	}
+
+	for _, u := range repoMDs {
+		if err := processRepoMD(u, localDir, verbose); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+type dbs struct {
+	primary, other string
+}
+
+func parse(localDir string) (CentOSPkgSet, error) {
+	fmt.Println("Processing sqlite files...")
+
+	dbPaths := make(map[string]dbs)
+	if err := filepath.WalkDir(localDir, func(path string, d fs.DirEntry, err error) error {
+		if err != nil {
+			return err
+		}
+		if d.IsDir() {
+			return nil
+		}
+		if !strings.HasSuffix(path, ".sqlite") {
+			return nil
+		}
+		dbp := dbPaths[filepath.Dir(path)]
+		if strings.HasSuffix(path, "-primary.sqlite") {
+			dbp.primary = path
+		} else if strings.HasSuffix(path, "-other.sqlite") {
+			dbp.other = path
+		}
+		dbPaths[filepath.Dir(path)] = dbp
+		return nil
+	}); err != nil {
+		return nil, err
+	}
+
+	allPkgs := make(CentOSPkgSet)
+	for _, db := range dbPaths {
+		pkgs, err := processSqlites(db)
+		if err != nil {
+			return nil, err
+		}
+		for pkg, cves := range pkgs {
+			for cve := range cves {
+				allPkgs.Add(pkg, cve)
+			}
+		}
+	}
+
+	return allPkgs, nil
+}
+
+func processRepoMD(mdURL url.URL, localDir string, verbose bool) error {
+	resp, err := http.Get(mdURL.String())
+	if err != nil {
+		return err
+	}
+	defer resp.Body.Close()
+
+	b, err := ioutil.ReadAll(resp.Body)
+	if err != nil {
+		return err
+	}
+	type location struct {
+		Href string `xml:"href,attr"`
+	}
+	type repoDataItem struct {
+		Type     string   `xml:"type,attr"`
+		Location location `xml:"location"`
+	}
+	type repoMetadata struct {
+		XMLName xml.Name       `xml:"repomd"`
+		Datas   []repoDataItem `xml:"data"`
+	}
+	var md repoMetadata
+	if err := xml.Unmarshal(b, &md); err != nil {
+		return err
+	}
+	for _, data := range md.Datas {
+		if data.Type != "primary_db" && data.Type != "other_db" {
+			continue
+		}
+		sqliteURL := mdURL
+		sqliteURL.Path = strings.TrimSuffix(sqliteURL.Path, "repomd.xml") + strings.TrimPrefix(data.Location.Href, "repodata/")
+		if err := downloadSqlite(sqliteURL, localDir, verbose); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func downloadSqlite(sqliteURL url.URL, localDir string, verbose bool) error {
```

Comment:
Could we use this as well in `SyncCPEDatabase`?

Reply:
Good idea. Done (added a new package `pkg/download` and used it on both the CPE and the CentOS generation).